src/linear_algebra/Hermitian_inner_product.lean

```diff
@@ -0,0 +1,669 @@
+
+noncomputable def inprod : α → α → ℂ := (herm_inner_product_space.to_sym_sesq_form α).to_sesq_form.sesq  
+
+infix `₀` : 74 := inprod 
+
+lemma conj_inprod (x y : α) : conj (x ₀ y) = y ₀ x := sym_sesq_form.map_sesq conj.ring_invo y x 
+
+lemma inprod_self_re_nonneg (x : α) : (x ₀ x).re ≥ 0 := sesq_self_re_nonneg x   
+
+lemma inprod_self_eq_zero_iff {x : α} : (inprod x x) = 0 ↔ x = (0 : α) := sesq_self_eq_zero_iff x
+
+@[simp] lemma inprod_add_left (x y z : α) : 
+(x + y) ₀ z = x ₀ z + y ₀ z := sesq_form.sesq_add_left _ _ _ _
+
+@[simp] lemma inprod_add_right (x y z : α) : 
+x ₀ (y + z) = x ₀ y + x ₀ z := sesq_form.sesq_add_right _ _ _ _
+
+@[simp] lemma inprod_smul_left (a : ℂ) (x y : α) :
+(a • x) ₀ y = a * (x ₀ y) := sesq_form.sesq_smul_left _ _ _ _
+
+@[simp] lemma inprod_smul_right (a : ℂ) (x y : α) :
+x ₀ (a • y) = conj(a) * (x ₀ y) := sesq_form.sesq_smul_right _ _ _ _
+
+open sym_sesq_form sesq_form
+
+@[simp] lemma zero_inprod (x : α) :
+0 ₀ x = 0 := zero_sesq x  
+
+@[simp] lemma inprod_zero (x : α) :
+x ₀ 0 = 0 := sesq_zero x 
+
+@[simp] lemma inprod_neg_left (x y : α) : 
+-x ₀ y = -(x ₀ y) := sesq_neg_left x y 
+
+@[simp] lemma inprod_neg_right (x y : α) : 
+x ₀ -y = -(x ₀ y) := sesq_neg_right x y   
+
+lemma inprod_sub_left (x y z : α) : 
+(x - y) ₀ z = (x ₀ z) - (y ₀ z) := sesq_sub_left x y z
+
+lemma inprod_sub_right (x y z : α) : 
+x ₀ (y - z) = (x ₀ y) - (x ₀ z) := sesq_sub_right x y z
+
+theorem inprod_self_im (x : α) :
+(x ₀ x).im = 0 :=
+(im_eq_zero_iff_conj_eq (x ₀ x)).mpr (conj_inprod x x)
+
+lemma inprod_self_re_eq_zero (x : α) : 
+(x ₀ x).re = 0 ↔ x = 0 := 
+begin
+  split; intros H,
+  { suffices : x ₀ x = 0,
+    { exact inprod_self_eq_zero_iff.mp this },
+    { apply complex.ext,
+      { simpa },
+
+      { rw [inprod_self_im, zero_im] }}},
+
+  { rw H,
+    simp }
+end
+
+lemma inprod_self_ne_zero_iff {x : α} : 
+(x ₀ x) ≠ 0 ↔ x ≠ 0 :=
+⟨ λ H, (iff_false_left H).mp inprod_self_eq_zero_iff, 
+  λ H, (iff_false_right H).mp inprod_self_eq_zero_iff⟩ 
+
+lemma inprod_self_re_ne_zero_iff {x : α} : 
+(x ₀ x).re ≠ 0 ↔ x ≠ 0 :=
+begin
+  split; intros H,
+  { have Ho : (x ₀ x) ≠ 0,
+    { intros Hx,
+      rw [Hx, zero_re] at H,
+      trivial },
+    exact inprod_self_ne_zero_iff.mp Ho },
+
+  { have Ho : (x ₀ x) ≠ 0,
+    { exact inprod_self_ne_zero_iff.mpr H },
+    exact ne_zero_im_zero_imp_re_ne_zero Ho (inprod_self_im x) } 
+end
+
+lemma inprod_re (x y : α) : (x ₀ y).re = (y ₀ x).re := 
+by rw [←conj_inprod, conj_re]
+
+lemma inprod_im (x y : α) : (x ₀ y).im = -(y ₀ x).im :=
+by rw [←conj_inprod, conj_im]
+
+noncomputable def herm_norm (x : α) := sqrt((x ₀ x).re)
+
+local notation |a| := herm_norm a 
+
+lemma mul_self_herm_norm (x : α) : |x| * |x| = (x ₀ x).re :=
+begin
+  dunfold herm_norm,
+  rw mul_self_sqrt (inprod_self_re_nonneg x),
+end
+
+lemma herm_norm_sqr (x : α) : |x|^2 = (x ₀ x).re := 
+by rw pow_two; exact mul_self_herm_norm x
+
+open classical
+
+theorem abs_inprod_le_mul_herm_norm (x y : α) :
```

Comment:
This proof (and the equality case further down)  cannot be that complicated. It needs rethinking. Maybe it should be proved later. A common issue we have with library building is that lemmas are ordered in our heads according to some idea of "being fundamental" which is not the actual logical order. Maybe you need to prove a bunch of trivial things first.

Reply:
I'll look into it, but this will probably take me some time. I have a feeling the issue is too many dunfolds followed by manipulation of square root terms that could be replaced by lemmas